File: MIDTERM-LAB-1/java/MonthlyPerformanceAnalyzer.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;

/**
 * MonthlyPerformanceAnalyzer.java
 *
 * Reads a VGChartz-style CSV dataset from a user-provided file path,
 * groups game records by release month, computes total sales per month,
 * displays a sorted monthly summary, and identifies the best-performing month.
 */
public class MonthlyPerformanceAnalyzer {

    /**
     * Prompts the user for a valid CSV file path.
     * Loops until a valid, readable CSV file is provided.
     *
     * @param scanner the Scanner for user input
     * @return the validated File object
     */
    private static File getValidFile(Scanner scanner) {
        File file;
        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = scanner.nextLine().trim();

            if (path.isEmpty()) {
                System.out.println("Error: File path cannot be empty. Please try again.");
                continue;
            }

            file = new File(path);

            if (!file.exists()) {
                System.out.println("Error: File does not exist. Please try again.");
                continue;
            }

            if (!file.isFile()) {
                System.out.println("Error: Path is not a file. Please try again.");
                continue;
            }

            if (!file.canRead()) {
                System.out.println("Error: File is not readable. Please try again.");
                continue;
            }

            if (!path.toLowerCase().endsWith(".csv")) {
                System.out.println("Error: File is not in CSV format (.csv). Please try again.");
                continue;
            }

            break;
        }
        return file;
    }

    /**
     * Parses the CSV file and returns a list of DataRecord objects.
     * Handles quoted fields that may contain commas.
     *
     * @param file the CSV file to parse
     * @return list of parsed DataRecord objects
     * @throws IOException if an I/O error occurs
     */
    private static List<DataRecord> loadDataset(File file) throws IOException {
        List<DataRecord> records = new ArrayList<>();
        int skippedRows = 0;

        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            String headerLine = reader.readLine();
            if (headerLine == null) {
                throw new IOException("CSV file is empty.");
            }

            // Validate that the CSV has the expected columns
            String[] headers = parseCsvLine(headerLine);
            int titleIndex = findColumnIndex(headers, "title");
            int totalSalesIndex = findColumnIndex(headers, "total_sales");
            int releaseDateIndex = findColumnIndex(headers, "release_date");

            if (titleIndex == -1 || totalSalesIndex == -1 || releaseDateIndex == -1) {
                throw new IOException(
                    "CSV file is missing required columns. Expected: title, total_sales, release_date. " +
                    "Found headers: " + headerLine
                );
            }

            String line;
            int lineNumber = 1;
            while ((line = reader.readLine()) != null) {
                lineNumber++;
                try {
                    String[] fields = parseCsvLine(line);

                    if (fields.length <= Math.max(titleIndex, Math.max(totalSalesIndex, releaseDateIndex))) {
                        skippedRows++;
                        continue;
                    }

                    String title = fields[titleIndex].trim();
                    String salesStr = fields[totalSalesIndex].trim();
                    String releaseDate = fields[releaseDateIndex].trim();

                    // Skip rows with empty sales or date
                    if (salesStr.isEmpty() || releaseDate.isEmpty()) {
                        skippedRows++;
                        continue;
                    }

                    double totalSales = Double.parseDouble(salesStr);
                    DataRecord record = new DataRecord(title, totalSales, releaseDate);

                    // Only include records with a valid month key
                    if (record.getMonthKey() != null) {
                        records.add(record);
                    } else {
                        skippedRows++;
                    }
                } catch (NumberFormatException e) {
                    skippedRows++;
                }
            }
        }

        System.out.println("Dataset loaded: " + records.size() + " valid records.");
        if (skippedRows > 0) {
            System.out.println("Skipped " + skippedRows + " rows (missing/invalid data).");
        }
        return records;
    }

    /**
     * Parses a single CSV line, handling quoted fields.
     *
     * @param line the CSV line to parse
     * @return array of field values
     */
    private static String[] parseCsvLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c == '"') {
                inQuotes = !inQuotes;
            } else if (c == ',' && !inQuotes) {
                fields.add(current.toString());
                current.setLength(0);
            } else {
                current.append(c);
            }
        }
        fields.add(current.toString());
        return fields.toArray(new String[0]);
    }

    /**
     * Finds the index of a column by name (case-insensitive).
     *
     * @param headers array of header names
     * @param name    the column name to find
     * @return the index, or -1 if not found
     */
    private static int findColumnIndex(String[] headers, String name) {
        for (int i = 0; i < headers.length; i++) {
            if (headers[i].trim().equalsIgnoreCase(name)) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Groups records by month and computes total sales per month.
     *
     * @param records the list of DataRecord objects
     * @return a TreeMap of month-key to total sales (sorted ascending)
     */
    private static TreeMap<String, Double> computeMonthlySales(List<DataRecord> records) {
        TreeMap<String, Double> monthlySales = new TreeMap<>();

        for (DataRecord record : records) {
            String monthKey = record.getMonthKey();
            if (monthKey != null) {
                monthlySales.merge(monthKey, record.getTotalSales(), Double::sum);
            }
        }
        return monthlySales;
    }

    /**
     * Displays the formatted monthly sales summary and identifies the best month.
     *
     * @param monthlySales the sorted map of monthly sales
     */
    private static void displayResults(TreeMap<String, Double> monthlySales) {
        if (monthlySales.isEmpty()) {
            System.out.println("No monthly data available to display.");
            return;
        }

        String bestMonth = null;
        double bestSales = Double.NEGATIVE_INFINITY;

        // Print header
        System.out.println();
        System.out.println("=======================================================");
        System.out.println("         MONTHLY PERFORMANCE SUMMARY                   ");
        System.out.println("=======================================================");
        System.out.printf("  %-12s | %18s | %8s%n", "Month", "Total Sales (M)", "Records");
        System.out.println("-------------------------------------------------------");

        // We also count records per month for informational purposes
        // but since we already aggregated, let's just show sales
        // For record counts, we'd need a second pass — let's track both
        // Actually, let's re-compute with counts
        // We already have totals; let's just display what we have.

        for (Map.Entry<String, Double> entry : monthlySales.entrySet()) {
            String month = entry.getKey();
            double sales = entry.getValue();

            System.out.printf("  %-12s | %18.2f |%n", month, sales);

            if (sales > bestSales) {
                bestSales = sales;
                bestMonth = month;
            }
        }

        System.out.println("-------------------------------------------------------");
        System.out.printf("  Total Months Analyzed: %d%n", monthlySales.size());
        System.out.println("=======================================================");
        System.out.println();
        System.out.println("*** BEST-PERFORMING MONTH ***");
        System.out.printf("  Month : %s%n", bestMonth);
        System.out.printf("  Sales : %.2fM%n", bestSales);
        System.out.println("=======================================================");
    }

    /**
     * Main entry point. Prompts user for file path, loads data, and displays results.
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("=======================================================");
        System.out.println("       MONTHLY PERFORMANCE ANALYZER                    ");
        System.out.println("=======================================================");
        System.out.println();

        // Step 1: Get valid file path from user
        File file = getValidFile(scanner);
        System.out.println("File found. Processing...");
        System.out.println();

        // Step 2: Load dataset
        List<DataRecord> records;
        try {
            records = loadDataset(file);
        } catch (IOException e) {
            System.out.println("Error reading dataset: " + e.getMessage());
            scanner.close();
            return;
        }

        if (records.isEmpty()) {
            System.out.println("No valid records found in the dataset.");
            scanner.close();
            return;
        }

        // Step 3: Compute monthly sales
        TreeMap<String, Double> monthlySales = computeMonthlySales(records);

        // Step 4: Display results
        displayResults(monthlySales);

        scanner.close();
    }
}

File: MIDTERM-LAB-1/javascript/MonthlyPerformanceAnalyzer.js
/**
 * MonthlyPerformanceAnalyzer.js
 *
 * Reads a VGChartz-style CSV dataset from a user-provided file path,
 * groups game records by release month, computes total sales per month,
 * displays a sorted monthly summary, and identifies the best-performing month.
 *
 * Usage: node MonthlyPerformanceAnalyzer.js
 */

const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// ─────────────────────────────────────────────
//  Data Record Module
// ─────────────────────────────────────────────

/**
 * Creates a DataRecord object from parsed CSV fields.
 * @param {string} title       - Game title
 * @param {number} totalSales  - Total worldwide sales (in millions)
 * @param {string} releaseDate - Release date in YYYY-MM-DD format
 * @returns {object} DataRecord with getMonthKey() method
 */
function createDataRecord(title, totalSales, releaseDate) {
    return {
        title: title,
        totalSales: totalSales,
        releaseDate: releaseDate,

        /**
         * Extracts the month key (YYYY-MM) from the release date.
         * @returns {string|null} month key or null if date is invalid
         */
        getMonthKey: function () {
            if (!releaseDate || releaseDate.length < 7) {
                return null;
            }
            return releaseDate.substring(0, 7);
        },

        toString: function () {
            const paddedTitle = title.substring(0, 50).padEnd(50);
            return `${paddedTitle} | Sales: ${totalSales.toFixed(2).padStart(8)}M | Date: ${releaseDate}`;
        }
    };
}

// ─────────────────────────────────────────────
//  CSV Parsing Module
// ─────────────────────────────────────────────

/**
 * Parses a single CSV line, handling quoted fields that may contain commas.
 * @param {string} line - A single CSV line
 * @returns {string[]} Array of field values
 */
function parseCsvLine(line) {
    const fields = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const c = line[i];
        if (c === '"') {
            inQuotes = !inQuotes;
        } else if (c === ',' && !inQuotes) {
            fields.push(current);
            current = '';
        } else {
            current += c;
        }
    }
    fields.push(current);
    return fields;
}

/**
 * Finds the index of a column by name (case-insensitive).
 * @param {string[]} headers - Array of header names
 * @param {string} name      - Column name to find
 * @returns {number} Column index, or -1 if not found
 */
function findColumnIndex(headers, name) {
    for (let i = 0; i < headers.length; i++) {
        if (headers[i].trim().toLowerCase() === name.toLowerCase()) {
            return i;
        }
    }
    return -1;
}

// ─────────────────────────────────────────────
//  File Loading Module
// ─────────────────────────────────────────────

/**
 * Loads and parses the CSV dataset from the given file path.
 * @param {string} filePath - Path to the CSV file
 * @returns {object[]} Array of DataRecord objects
 */
function loadDataset(filePath) {
    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split(/\r?\n/);

    if (lines.length === 0 || lines[0].trim() === '') {
        throw new Error('CSV file is empty.');
    }

    const headers = parseCsvLine(lines[0]);
    const titleIndex = findColumnIndex(headers, 'title');
    const totalSalesIndex = findColumnIndex(headers, 'total_sales');
    const releaseDateIndex = findColumnIndex(headers, 'release_date');

    if (titleIndex === -1 || totalSalesIndex === -1 || releaseDateIndex === -1) {
        throw new Error(
            'CSV file is missing required columns. Expected: title, total_sales, release_date. ' +
            'Found headers: ' + lines[0]
        );
    }

    const records = [];
    let skippedRows = 0;
    const maxIndex = Math.max(titleIndex, totalSalesIndex, releaseDateIndex);

    for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (line === '') {
            continue;
        }

        try {
            const fields = parseCsvLine(line);

            if (fields.length <= maxIndex) {
                skippedRows++;
                continue;
            }

            const title = fields[titleIndex].trim();
            const salesStr = fields[totalSalesIndex].trim();
            const releaseDate = fields[releaseDateIndex].trim();

            // Skip rows with empty sales or date
            if (salesStr === '' || releaseDate === '') {
                skippedRows++;
                continue;
            }

            const totalSales = parseFloat(salesStr);
            if (isNaN(totalSales)) {
                skippedRows++;
                continue;
            }

            const record = createDataRecord(title, totalSales, releaseDate);

            if (record.getMonthKey() !== null) {
                records.push(record);
            } else {
                skippedRows++;
            }
        } catch (e) {
            skippedRows++;
        }
    }

    console.log(`Dataset loaded: ${records.length} valid records.`);
    if (skippedRows > 0) {
        console.log(`Skipped ${skippedRows} rows (missing/invalid data).`);
    }

    return records;
}

// ─────────────────────────────────────────────
//  Analytics Module
// ─────────────────────────────────────────────

/**
 * Groups records by month and computes total sales per month.
 * @param {object[]} records - Array of DataRecord objects
 * @returns {Map<string, number>} Map of month-key to total sales
 */
function computeMonthlySales(records) {
    const monthlySales = new Map();

    for (const record of records) {
        const monthKey = record.getMonthKey();
        if (monthKey !== null) {
            const current = monthlySales.get(monthKey) || 0;
            monthlySales.set(monthKey, current + record.totalSales);
        }
    }

    return monthlySales;
}

/**
 * Sorts monthly sales by month key in ascending order.
 * @param {Map<string, number>} monthlySales - Unsorted monthly sales map
 * @returns {Array<[string, number]>} Sorted array of [month, sales] pairs
 */
function sortMonthlySales(monthlySales) {
    const entries = Array.from(monthlySales.entries());
    entries.sort((a, b) => a[0].localeCompare(b[0]));
    return entries;
}

// ─────────────────────────────────────────────
//  Display Module
// ─────────────────────────────────────────────

/**
 * Displays the formatted monthly sales summary and identifies the best month.
 * @param {Array<[string, number]>} sortedMonthlySales - Sorted monthly sales data
 */
function displayResults(sortedMonthlySales) {
    if (sortedMonthlySales.length === 0) {
        console.log('No monthly data available to display.');
        return;
    }

    let bestMonth = null;
    let bestSales = -Infinity;

    console.log('');
    console.log('=======================================================');
    console.log('         MONTHLY PERFORMANCE SUMMARY                   ');
    console.log('=======================================================');
    console.log(`  ${'Month'.padEnd(12)} | ${'Total Sales (M)'.padStart(18)} `);
    console.log('-------------------------------------------------------');

    for (const [month, sales] of sortedMonthlySales) {
        console.log(`  ${month.padEnd(12)} | ${sales.toFixed(2).padStart(18)} `);

        if (sales > bestSales) {
            bestSales = sales;
            bestMonth = month;
        }
    }

    console.log('-------------------------------------------------------');
    console.log(`  Total Months Analyzed: ${sortedMonthlySales.length}`);
    console.log('=======================================================');
    console.log('');
    console.log('*** BEST-PERFORMING MONTH ***');
    console.log(`  Month : ${bestMonth}`);
    console.log(`  Sales : ${bestSales.toFixed(2)}M`);
    console.log('=======================================================');
}

// ─────────────────────────────────────────────
//  File Path Validation & Main Flow
// ─────────────────────────────────────────────

/**
 * Validates that the given path points to a readable CSV file.
 * @param {string} filePath - The file path to validate
 * @returns {string|null} Error message, or null if valid
 */
function validateFilePath(filePath) {
    if (!filePath || filePath.trim() === '') {
        return 'Error: File path cannot be empty. Please try again.';
    }

    filePath = filePath.trim();

    if (!fs.existsSync(filePath)) {
        return 'Error: File does not exist. Please try again.';
    }

    try {
        const stats = fs.statSync(filePath);
        if (!stats.isFile()) {
            return 'Error: Path is not a file. Please try again.';
        }
    } catch (e) {
        return 'Error: Cannot access file. Please try again.';
    }

    try {
        fs.accessSync(filePath, fs.constants.R_OK);
    } catch (e) {
        return 'Error: File is not readable. Please try again.';
    }

    if (!filePath.toLowerCase().endsWith('.csv')) {
        return 'Error: File is not in CSV format (.csv). Please try again.';
    }

    return null;
}

/**
 * Asks the user for a valid file path, looping until one is provided.
 * Then processes the dataset and displays results.
 */
function askFilePath() {
    rl.question('Enter dataset file path: ', function (path) {
        const error = validateFilePath(path);

        if (error) {
            console.log(error);
            askFilePath();
            return;
        }

        console.log('File found. Processing...');
        console.log('');

        try {
            // Step 1: Load dataset
            const records = loadDataset(path.trim());

            if (records.length === 0) {
                console.log('No valid records found in the dataset.');
                rl.close();
                return;
            }

            // Step 2: Compute monthly sales
            const monthlySales = computeMonthlySales(records);

            // Step 3: Sort by month ascending
            const sortedMonthlySales = sortMonthlySales(monthlySales);

            // Step 4: Display results
            displayResults(sortedMonthlySales);
        } catch (e) {
            console.log('Error reading dataset: ' + e.message);
        }

        rl.close();
    });
}

// Main execution
console.log('=======================================================');
console.log('       MONTHLY PERFORMANCE ANALYZER                    ');
console.log('=======================================================');
console.log('');

askFilePath();
